[ADR] New Authorization roles (#2677)
epinio/epinio: docs/decisions/20231013-authorization-roles.md

---
status: {proposed}
date: {2023-10-13}
---

# Authorization Roles

## Authorization

Until version `v1.10.0` Epinio has a very simple authorization mechanism. A user can be `admin` or `user`.  
An admin can perform everything, while a user can do anything but only in the namespaces he created or where has permissions.

This was pretty simple to implement but not very flexible.

## Decision Drivers

A new more flexible authorization was needed (https://github.com/epinio/epinio/issues/1946).

Some of the requirements:
- Be easy to understand and administer
- Provide least privileges
- Allow separate roles for listing, reading details, editing, and creating
- Allow soft multi-tenancy across namespaces (e.g. user A has read write access for apps on one namespace but read only on another namespace)
- Not require too many changes to the existing code base

## Considered Options

- Secret with whitelist ACL per namespace
- Role Bindings as their own CRDs
- ConfigMaps defined Roles with predefined Actions

## Decision Outcome

> Chosen option 3 (__"ConfigMaps defined Roles with predefined Actions"__). It was the easier and more flexible solution. It will not require many changes in the code, and it will be easy to customize and implement.

Reviewer: ```suggestion
Chosen option 3 (__"ConfigMaps defined Roles with predefined Actions"__). It is believed to be the easiest and most flexible solution. It will not require many changes in the code, and it will be easy to customize and implement.
```

We could also define some basic roles that can be used.

## Pros and Cons of the Options

### Secret with whitelist ACL per namespace

A secret per namespace containing the whitelist ACL could provide this without too much administrative overhead.

The ACL data in the secret could look something like this

```yaml
application:
  jwt_claim_name: rwlcd
  jwt_claim2: l
service:
  jwt_claim_name: rw
  jwt_claim2: rld
```

Where the RHS of the claim mapping is a string with a combination of letters:

```
r -- read
w -- write/edit
l -- list
c -- create
d -- delete
a -- all
```

A user with multiple claims would get access to the union of access and an admin would have access to everything.

Additionally, as long as the secret/configmap is mounted in the pod without any subpath, changing it would not require a restart of the API.

#### Pros

* Clear "resource" and "permission" description

#### Cons

> * Not sure how to handle newly created namespace

Reviewer: ```suggestion
* Not sure how to handle newly created namespaces
```

> * Some operations are not strictly tight to a "resource". How to avoid the `exec` but allowing the creation of apps?

Reviewer: ```suggestion
* Some operations are not strictly tied to a "resource". How to avoid the `exec` but allowing the creation of apps?
```

### Role Bindings as their own CRDs

Building custom Role Bindings as their own CRDs.

#### Pros

* Clear and structured implementation

#### Cons

* Difficult/complex to implement and evolve
* Complex to manage

### ConfigMaps defined Roles with predefined Actions

> 

Reviewer: Generally speaking this part needs quite a bit more elaboration of syntax and semantics of the config maps, the action.yaml, the actions (if there is a difference).

Treat this part like writing a specification with __all__ the details.
Currently a lot seems to be missing and glossed over.

> To avoid changing too much the current implementation we can add "roles" to an Epinio user. At the moment the user can only have one role, but this can be extended with a list instead of a fixed value. Also the role can be namescoped adding a `::` delimiter.

Reviewer: ```suggestion
To avoid changing the current implementation too much we can add "roles" to an Epinio user. While a user can only have one role at the moment, this can be extended in the future with a list instead of a single value. Furthermore a role can be namespace-scoped by adding a `::` delimiter.
```

Reviewer: Question here. Why `::` for the delimiter ? What other delimiters were considered ? (`/`, `:`, `,`, ...).

Author: I was looking for a character that was not "reserved", or conflicting with the Kubernetes rules (btw annotations are more flexible I think). Commas are usually used to separate completely different "set", and it was already used to separate different roles. I was looking for something that seemed to be like a URN.

While testing the OIDC authentication with Github the returned groups are formed with the `org:team` pair, so it looked ok to have a similar (but different) delimiter for `role::namespace`. Also `/` could have been fine btw, but the roles annotations would have looked "weird". Checking the yaml basically the choices would have been:

```yaml
metadata:
  annotations:
    epinio.io/roles: "admin,user::workspace,reader::foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin,user/workspace,reader/foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin,user:workspace,reader:foobar"
```

This is the outcome of the `rolesMapping` section in the dex.yaml configuration:

```yaml
    - connectorId: github
      groups:
        - id: epinio:epinio
          roles:
            - admin
        - id: epinio:ui
          roles:
            - user::workspace
        - id: Org1:TeamBlue
          roles:
            - user
            - admin::workspace
```

I'm ok with changing it to something different. Maybe the single `:` (or triple? `:::`) , or the `/` also looks nice to me.

Regarding delimiters I was not sure also about the comma. Maybe a pipe `|` could have been an option easier to read:

```yaml
metadata:
  annotations:
    epinio.io/roles: "admin|user::workspace|reader::foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin|user:workspace|reader:foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin|user/workspace|reader/foobar"
```

Reviewer: I am ok with the comma for role separation. I would have used `:` or `/` for the namespace separation.

Author: Ok, I think I will change it to a single `:` then. :+1:

The Secret with the definition of the User can contain an annotation with the assigned roles.

```yaml
metadata:
  annotations:
    epinio.io/roles: epinio-role-reader,admin::workspace
  labels:
    epinio.io/role: user
```

*Note:*

> an annotation will be used because labels have some limitations about the lenght and we don't want to be restricted about the number of roles. Also we don't need to perform any kind of lookup over the role.

Reviewer: ```suggestion
An annotation will be used because labels have length-limitations and we don't want to be restricted about the number of roles. Also we don't need to perform any kind of lookup over the role.
```

> In the previous example the user will have the `epinio-role-reader` and `admin::workspace` roles. When working in the workspace namespace the `admin` role will be used, otherwise the `epinio-role-reader` will take over.

Reviewer: ```suggestion
In the previous example the user has the roles `epinio-role-reader` and `admin::workspace`. When working in the  namespace `workspace` the `admin` role will be used, otherwise the role `epinio-role-reader`.
```

> A Role can be simply defined as a ConfigMap (no need to be a secret) with a special label (`epinio.io/role: "true"`). This role contains the `actions` that this role can perform, and some metadata that can be used as descriptions (i.e.: the `name`):

Reviewer: ```suggestion
A `Role` is defined as a kubernetes `ConfigMap` (no need to be a `Secret`) with a special label (`epinio.io/role: "true"`). This resource contains the `actions` that the role can perform, and some metadata that can be used for descriptions (i.e.: the `name`):
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    epinio.io/role: "true"
  name: epinio-role-namespace
  namespace: epinio
data:
  name: Epinio Namespace Role
  actions: |
    namespace
```

> This actions could be **hardcoded** in a yaml file. This simplify the management of them, and also the flexibility. An action can have some "dependencies", i.e.: the `namespace` action is a union of the `namespace_show`, `namespace_delete`, and so on.

Reviewer: ```suggestion
The actions are **hardcoded** as an embedded yaml file. This simplifies their management, and also enhances flexibility. An action can have some "dependencies", i.e.: the `namespace` action is a union of the `namespace_show`, `namespace_delete`, and so on.
```

> Every action will have a set of endpoints that will allow. Some Epinio operations are formed by multiple endpoints, i.e. the `app_push` consists of a Create, Update and others.

Reviewer: ```suggestion
Every action lists the set of endpoints it allows. Note that some Epinio operations are formed from multiple endpoints, i.e. the `app_push` consists of a Create, Update and others.
```

| `configuration_read`  | Read permissions (list, show)
| `configuration_write` | Write permissions (create, delete)<br/>Depends on: `configuration_read`
| `configuration`       | All the above<br/>Depends on: `configuration_read`, `configuration_write`

##### Service

These actions enable operations on Service commands and resources. 

| Action ID             | Description 
|-----------------------|-------------
| `service_read`        | Read permissions (list, show)
| `service_write`       | Write permissions (create, delete, bind, unbind)<br/>Depends on: `service_read`
| `service_portforward` | Open a tunnel with the `port-forward` command
| `service`             | All the above<br/>Depends on: `service_read`, `service_write`, `service_portforward`

##### Gitconfig

These actions enable operations on Gitconfig commands and resources.

| Action ID         | Description 
|-------------------|-------------
| `gitconfig_read`    | Read permissions (list, show)
| `gitconfig_write`   | Write permissions (create, delete)<br/>Depends on: `gitconfig_read`
| `gitconfig`         | All the above<br/>Depends on: `gitconfig_read`, `gitconfig_write`

##### Export Registries

> This action enable operations on Export Registries commands and resources. Only read operations are available.

Reviewer: ```suggestion
This action enables operations on Export Registries commands and resources. Only read operations are available.
```
